Kafka Client does not support SASL with JASS

**Describe the bug**

Kafka SASL is not supported in native mode.

Even with the added set of classes, it fails with:


```
Caused by: javax.security.sasl.SaslException: Cannot get userid/password [Caused by javax.security.auth.callback.UnsupportedCallbackException: Could not login: the client is being asked for a password, but the Kafka client code does not currently support obtaining a password from the user.]
 at com.sun.security.sasl.ClientFactoryImpl.getUserInfo(ClientFactoryImpl.java:157)
        at com.sun.security.sasl.ClientFactoryImpl.createSaslClient(ClientFactoryImpl.java:94)
        at javax.security.sasl.Sasl.createSaslClient(Sasl.java:433)
        at org.apache.kafka.common.security.authenticator.SaslClientAuthenticator.lambda$createSaslClient$0(SaslClientAuthenticator.java:183)
        at java.security.AccessController.doPrivileged(AccessController.java:144)
        at javax.security.auth.Subject.doAs(Subject.java:423)
```


**Current Status**

I have added the following code to `io.quarkus.kafka.client.deployment.KafkaProcessor`.

```
reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, SaslClientCallbackHandler.class.getName()));
        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, DefaultLogin.class.getName()));
        reflectiveClass
                .produce(new ReflectiveClassBuildItem(true, false,
                        AbstractLogin.DefaultLoginCallbackHandler.class.getName()));
        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, PlainLoginModule.class.getName()));
```

--- a/extensions/kafka-client/deployment/pom.xml
+++ b/extensions/kafka-client/deployment/pom.xml
@@ -31,6 +31,10 @@
             <artifactId>quarkus-junit5-internal</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc-deployment</artifactId>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
@@ -4,12 +4,17 @@ import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
 
+import javax.security.auth.spi.LoginModule;
+
 import org.apache.kafka.clients.consumer.RangeAssignor;
 import org.apache.kafka.clients.consumer.RoundRobinAssignor;
 import org.apache.kafka.clients.consumer.StickyAssignor;
 import org.apache.kafka.clients.consumer.internals.PartitionAssignor;
 import org.apache.kafka.clients.producer.Partitioner;
 import org.apache.kafka.clients.producer.internals.DefaultPartitioner;
+import org.apache.kafka.common.security.authenticator.AbstractLogin;
+import org.apache.kafka.common.security.authenticator.DefaultLogin;
+import org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler;
 import org.apache.kafka.common.serialization.ByteArrayDeserializer;
 import org.apache.kafka.common.serialization.ByteArraySerializer;
 import org.apache.kafka.common.serialization.ByteBufferDeserializer;
@@ -32,12 +37,18 @@ import org.apache.kafka.common.serialization.StringDeserializer;
 import org.apache.kafka.common.serialization.StringSerializer;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
+import org.jboss.jandex.Type;
+import org.jboss.jandex.Type.Kind;
 
+import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.deployment.Capabilities;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.builditem.IndexDependencyBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
+import io.quarkus.kafka.client.runtime.KafkaRuntimeConfigProducer;
 import io.quarkus.kafka.client.serialization.JsonbDeserializer;
 import io.quarkus.kafka.client.serialization.JsonbSerializer;
 import io.quarkus.kafka.client.serialization.ObjectMapperDeserializer;
@@ -67,7 +78,7 @@ public class KafkaProcessor {
             IntegerDeserializer.class,
             ByteBufferDeserializer.class,
             StringDeserializer.class,
-            FloatDeserializer.class,
+            FloatDeserializer.class
     };
 
     @BuildStep
@@ -109,6 +120,31 @@ public class KafkaProcessor {
         // classes needed to perform reflection on DirectByteBuffer - only really needed for Java 8
         reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, "java.nio.DirectByteBuffer"));
         reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, "sun.misc.Cleaner"));
+    }
+
+    @BuildStep
+    public AdditionalBeanBuildItem runtimeConfig() {
+        return AdditionalBeanBuildItem.builder()
+                .addBeanClass(KafkaRuntimeConfigProducer.class)
+                .setUnremovable()
+                .build();
+    }
+
+    @BuildStep
+    public void withSasl(BuildProducer<IndexDependencyBuildItem> indexDependency,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchy) {
+
+        indexDependency.produce(new IndexDependencyBuildItem("org.apache.kafka", "kafka-clients"));
+
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, AbstractLogin.DefaultLoginCallbackHandler.class));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, SaslClientCallbackHandler.class));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, DefaultLogin.class));
+
+        final Type loginModuleType = Type
+                .create(DotName.createSimple(LoginModule.class.getCanonicalName()), Kind.CLASS);
+
+        reflectiveHierarchy.produce(new ReflectiveHierarchyBuildItem(loginModuleType));
     }
 
     private static void collectImplementors(Set<DotName> set, CombinedIndexBuildItem indexBuildItem, Class<?> cls) {

--- a/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java
+++ b/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java
@@ -6,11 +6,12 @@ import java.util.Map;
 import javax.annotation.PostConstruct;
 import javax.annotation.PreDestroy;
 import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.inject.Named;
 
 import org.apache.kafka.clients.admin.AdminClient;
 import org.apache.kafka.clients.admin.AdminClientConfig;
 import org.apache.kafka.common.Node;
-import org.eclipse.microprofile.config.inject.ConfigProperty;
 import org.eclipse.microprofile.health.HealthCheck;
 import org.eclipse.microprofile.health.HealthCheckResponse;
 import org.eclipse.microprofile.health.HealthCheckResponseBuilder;
@@ -20,15 +21,15 @@ import org.eclipse.microprofile.health.Readiness;
 @ApplicationScoped
 public class KafkaHealthCheck implements HealthCheck {
 
-    @ConfigProperty(name = "kafka.bootstrap.servers", defaultValue = "localhost:9092")
-    private String bootstrapServers;
+    @Inject
+    @Named("default-kafka-broker")
+    Map<String, Object> config;
 
     private AdminClient client;
 
     @PostConstruct
     void init() {
-        Map<String, Object> conf = new HashMap<>();
-        conf.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        Map<String, Object> conf = new HashMap<>(config);
         conf.put(AdminClientConfig.REQUEST_TIMEOUT_MS_CONFIG, "5000");
         client = AdminClient.create(conf);
     }

--- a/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java
+++ b/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java
@@ -1,0 +1,45 @@
+package io.quarkus.kafka.client.runtime;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.StreamSupport;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Produces;
+import javax.inject.Named;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+
+import io.quarkus.arc.DefaultBean;
+
+@Dependent
+public class KafkaRuntimeConfigProducer {
+
+    private String configPrefix = "kafka";
+
+    @Produces
+    @DefaultBean
+    @ApplicationScoped
+    @Named("default-kafka-broker")
+    public Map<String, Object> createKafkaRuntimeConfig() {
+        Map<String, Object> properties = new HashMap<>();
+        final Config config = ConfigProvider.getConfig();
+
+        StreamSupport
+                .stream(config.getPropertyNames().spliterator(), false)
+                .map(String::toLowerCase)
+                .filter(name -> name.startsWith(configPrefix))
+                .distinct()
+                .sorted()
+                .forEach(name -> {
+                    final String key = name.substring(configPrefix.length() + 1).toLowerCase().replaceAll("[^a-z0-9.]", ".");
+                    final String value = config.getOptionalValue(name, String.class).orElse("");
+                    properties.put(key, value);
+                });
+
+        return properties;
+    }
+
+}

--- a/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/graal/SaslClientCallbackHandlerSubstitution.java
+++ b/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/graal/SaslClientCallbackHandlerSubstitution.java
@@ -1,0 +1,131 @@
+package io.quarkus.kafka.client.runtime.graal;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.security.auth.Subject;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.login.AppConfigurationEntry;
+import javax.security.auth.spi.LoginModule;
+import javax.security.sasl.AuthorizeCallback;
+import javax.security.sasl.RealmCallback;
+
+import org.apache.kafka.common.config.SaslConfigs;
+import org.apache.kafka.common.security.auth.AuthenticateCallbackHandler;
+import org.apache.kafka.common.security.auth.SaslExtensions;
+import org.apache.kafka.common.security.auth.SaslExtensionsCallback;
+import org.apache.kafka.common.security.authenticator.SaslClientCallbackHandler;
+import org.apache.kafka.common.security.scram.ScramExtensionsCallback;
+import org.apache.kafka.common.security.scram.internals.ScramMechanism;
+import org.apache.kafka.common.utils.Utils;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.Inject;
+import com.oracle.svm.core.annotate.KeepOriginal;
+import com.oracle.svm.core.annotate.RecomputeFieldValue;
+import com.oracle.svm.core.annotate.RecomputeFieldValue.Kind;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+@TargetClass(value = SaslClientCallbackHandler.class)
+final class Target_org_apache_kafka_common_security_authenticator_SaslClientCallbackHandler
+        implements AuthenticateCallbackHandler {
+
+    @Alias
+    @RecomputeFieldValue(kind = Kind.None)
+    String mechanism;
+
+    @Inject
+    @RecomputeFieldValue(kind = Kind.None)
+    Subject subject;
+
+    @Inject
+    @RecomputeFieldValue(kind = Kind.None)
+    Map<String, ?> sharedState;
+
+    @Override
+    @Substitute
+    public void configure(Map<String, ?> configs, String saslMechanism, List<AppConfigurationEntry> jaasConfigEntries) {
+        mechanism = saslMechanism;
+
+        for (AppConfigurationEntry entry : jaasConfigEntries) {
+            try {
+                if (subject == null) {
+                    subject = new Subject();
+                }
+                if (sharedState == null) {
+                    sharedState = new HashMap<>();
+                }
+                LoginModule result = Utils.newInstance(entry.getLoginModuleName(), LoginModule.class);
+                result.initialize(subject, this, sharedState, entry.getOptions());
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Override
+    @Substitute
+    public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
+        for (Callback callback : callbacks) {
+            if (callback instanceof NameCallback) {
+                NameCallback nc = (NameCallback) callback;
+                if (subject != null && !subject.getPublicCredentials(String.class).isEmpty()) {
+                    nc.setName(subject.getPublicCredentials(String.class).iterator().next());
+                } else {
+                    nc.setName(nc.getDefaultName());
+                }
+            } else if (callback instanceof PasswordCallback) {
+                if (subject != null && !subject.getPrivateCredentials(String.class).isEmpty()) {
+                    char[] password = subject.getPrivateCredentials(String.class).iterator().next()
+                            .toCharArray();
+                    ((PasswordCallback) callback).setPassword(password);
+                } else {
+                    String errorMessage = "Could not login: the client is being asked for a password, but the Kafka" +
+                            " client code does not currently support obtaining a password from the user.";
+                    throw new UnsupportedCallbackException(callback, errorMessage);
+                }
+            } else if (callback instanceof RealmCallback) {
+                RealmCallback rc = (RealmCallback) callback;
+                rc.setText(rc.getDefaultText());
+            } else if (callback instanceof AuthorizeCallback) {
+                AuthorizeCallback ac = (AuthorizeCallback) callback;
+                String authId = ac.getAuthenticationID();
+                String authzId = ac.getAuthorizationID();
+                ac.setAuthorized(authId.equals(authzId));
+                if (ac.isAuthorized()) {
+                    ac.setAuthorizedID(authzId);
+                }
+            } else if (callback instanceof ScramExtensionsCallback) {
+                if (ScramMechanism.isScram(mechanism) && subject != null && !subject
+                        .getPublicCredentials(Map.class).isEmpty()) {
+                    @SuppressWarnings("unchecked")
+                    Map<String, String> extensions = (Map<String, String>) subject
+                            .getPublicCredentials(Map.class).iterator().next();
+                    ((ScramExtensionsCallback) callback).extensions(extensions);
+                }
+            } else if (callback instanceof SaslExtensionsCallback) {
+                if (!SaslConfigs.GSSAPI_MECHANISM.equals(mechanism) &&
+                        subject != null && !subject.getPublicCredentials(SaslExtensions.class).isEmpty()) {
+                    SaslExtensions extensions = subject.getPublicCredentials(SaslExtensions.class).iterator()
+                            .next();
+                    ((SaslExtensionsCallback) callback).extensions(extensions);
+                }
+            } else {
+                throw new UnsupportedCallbackException(callback, "Unrecognized SASL ClientCallback");
+            }
+        }
+    }
+
+    @Override
+    @Substitute
+    @KeepOriginal
+    public void close() {
+
+    }
+
+}

--- a/integration-tests/kafka/src/main/java/io/quarkus/it/kafka/sasl/SaslKafkaEndpoint.java
+++ b/integration-tests/kafka/src/main/java/io/quarkus/it/kafka/sasl/SaslKafkaEndpoint.java
@@ -1,0 +1,64 @@
+package io.quarkus.it.kafka.sasl;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Properties;
+
+import javax.annotation.PostConstruct;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.config.SaslConfigs;
+import org.apache.kafka.common.serialization.IntegerDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+/**
+ * Endpoint to check the SSL/SASL connection.
+ */
+@Path("/sasl")
+public class SaslKafkaEndpoint {
+
+    private Consumer<Integer, String> consumer;
+
+    @PostConstruct
+    public void create() {
+        consumer = createConsumer();
+    }
+
+    @GET
+    public String get() {
+        final ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofMillis(60000));
+        if (records.isEmpty()) {
+            return null;
+        }
+        return records.iterator().next().value();
+    }
+
+    private static void addJAAS(Properties props) {
+        props.setProperty(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, "SASL_PLAINTEXT");
+        props.setProperty(SaslConfigs.SASL_MECHANISM, "PLAIN");
+        props.setProperty(SaslConfigs.SASL_JAAS_CONFIG,
+                "org.apache.kafka.common.security.plain.PlainLoginModule required "
+                        + "username=\"client\" "
+                        + "password=\"client-secret\";");
+    }
+
+    public static KafkaConsumer<Integer, String> createConsumer() {
+        Properties props = new Properties();
+        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:19094");
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, "test-consumer");
+        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, IntegerDeserializer.class.getName());
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        addJAAS(props);
+        KafkaConsumer<Integer, String> consumer = new KafkaConsumer<>(props);
+        consumer.subscribe(Collections.singletonList("test-sasl-consumer"));
+        return consumer;
+    }
+}

--- a/integration-tests/kafka/src/main/java/io/quarkus/it/kafka/ssl/SslKafkaEndpoint.java
+++ b/integration-tests/kafka/src/main/java/io/quarkus/it/kafka/ssl/SslKafkaEndpoint.java
@@ -1,4 +1,4 @@
-package io.quarkus.it.kafka.sasl;
+package io.quarkus.it.kafka.ssl;
 
 import java.io.File;
 import java.time.Duration;
@@ -19,10 +19,10 @@ import org.apache.kafka.common.serialization.IntegerDeserializer;
 import org.apache.kafka.common.serialization.StringDeserializer;
 
 /**
- * Endpoint to check the SSL/SASL connection.
+ * Endpoint to check the SSL connection.
  */
-@Path("/sasl")
-public class SaslKafkaEndpoint {
+@Path("/ssl")
+public class SslKafkaEndpoint {
 
     private Consumer<Integer, String> consumer;
 

--- a/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/KafkaSASLTestResource.java
+++ b/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/KafkaSASLTestResource.java
@@ -1,0 +1,87 @@
+package io.quarkus.it.kafka;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Properties;
+
+import io.debezium.kafka.KafkaCluster;
+import io.debezium.util.Testing;
+import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
+
+public class KafkaSASLTestResource implements QuarkusTestResourceLifecycleManager {
+
+    private KafkaCluster kafka;
+
+    @Override
+    public Map<String, String> start() {
+        try {
+            File directory = Testing.Files.createTestingDirectory("kafka-data-sasl", true);
+
+            enableServerJaasConf();
+
+            Properties props = new Properties();
+            props.setProperty("zookeeper.connection.timeout.ms", "45000");
+            props.setProperty("listener.security.protocol.map", "CLIENT:SASL_PLAINTEXT");
+            props.setProperty("listeners", "CLIENT://:19094");
+            props.setProperty("inter.broker.listener.name", "CLIENT");
+
+            props.setProperty("sasl.enabled.mechanisms", "PLAIN");
+            props.setProperty("sasl.mechanism.inter.broker.protocol", "PLAIN");
+
+            kafka = new KafkaCluster()
+                    .withPorts(2184, 19094)
+                    .addBrokers(1)
+                    .usingDirectory(directory)
+                    .deleteDataUponShutdown(true)
+                    .withKafkaConfiguration(props)
+                    .deleteDataPriorToStartup(true)
+                    .startup();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        return Collections.emptyMap();
+    }
+
+    @Override
+    public void stop() {
+        if (kafka != null) {
+            kafka.shutdown();
+        }
+    }
+
+    public static void enableServerJaasConf() throws IOException {
+        final Path conf = Files.createTempFile("kafka-server-jaas.", ".conf");
+        String serverConfiguration = "KafkaServer { "
+                + "org.apache.kafka.common.security.plain.PlainLoginModule required "
+                + "username=\"broker\" "
+                + "password=\"broker-secret\" "
+                + "user_broker=\"broker-secret\" "
+                + "user_client=\"client-secret\"; };";
+
+        Files.write(conf, ("client." + serverConfiguration).getBytes(StandardCharsets.UTF_8),
+                StandardOpenOption.CREATE);
+
+        System.setProperty("java.security.auth.login.config", conf.toAbsolutePath().toString());
+        System.setProperty("zookeeper.sasl.client", "false");
+    }
+
+    /**
+     * Make sure this runs first otherwise system property {@code java.security.auth.login.config}
+     * is ignored since {@link javax.security.auth.login.Configuration#configuration}
+     * is already initialized with default, empty configuration.
+     *
+     * @return {@link Integer#MIN_VALUE}
+     */
+    @Override
+    public int order() {
+        return Integer.MIN_VALUE;
+    }
+
+}

--- a/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/KafkaSSLTestResource.java
+++ b/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/KafkaSSLTestResource.java
@@ -16,14 +16,14 @@ import io.debezium.kafka.KafkaCluster;
 import io.debezium.util.Testing;
 import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
 
-public class KafkaSASLTestResource implements QuarkusTestResourceLifecycleManager {
+public class KafkaSSLTestResource implements QuarkusTestResourceLifecycleManager {
 
     private KafkaCluster kafka;
 
     @Override
     public Map<String, String> start() {
         try {
-            File directory = Testing.Files.createTestingDirectory("sasl-kafka-data", true);
+            File directory = Testing.Files.createTestingDirectory("kafka-data-ssl", true);
             File sslDir = sslDir(directory, true);
 
             Path ksPath = new File(sslDir, "kafka-keystore.p12").toPath();
@@ -82,7 +82,7 @@ public class KafkaSASLTestResource implements QuarkusTestResourceLifecycleManage
 
     public static File sslDir(File directory, boolean removeExistingContent) throws IOException {
         if (directory == null) {
-            directory = Testing.Files.createTestingDirectory("sasl-kafka-data", removeExistingContent);
+            directory = Testing.Files.createTestingDirectory("kafka-data-ssl", removeExistingContent);
         }
 
         File targetDir = directory.getParentFile().getParentFile();

--- a/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/SaslKafkaConsumerTest.java
+++ b/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/SaslKafkaConsumerTest.java
@@ -1,0 +1,52 @@
+package io.quarkus.it.kafka;
+
+import java.util.Properties;
+
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.config.SaslConfigs;
+import org.apache.kafka.common.serialization.IntegerSerializer;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.common.QuarkusTestResource;
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+
+@QuarkusTest
+@QuarkusTestResource(KafkaSASLTestResource.class)
+public class SaslKafkaConsumerTest {
+
+    private static void addJaas(Properties props) {
+        props.setProperty(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, "SASL_PLAINTEXT");
+        props.setProperty(SaslConfigs.SASL_MECHANISM, "PLAIN");
+        props.setProperty(SaslConfigs.SASL_JAAS_CONFIG,
+                "org.apache.kafka.common.security.plain.PlainLoginModule required "
+                        + "username=\"client\" "
+                        + "password=\"client-secret\";");
+    }
+
+    public static Producer<Integer, String> createProducer() {
+        Properties props = new Properties();
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:19094");
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "sasl-test-producer");
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class.getName());
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        addJaas(props);
+
+        return new KafkaProducer<>(props);
+    }
+
+    @Test
+    public void testReception() {
+        Producer<Integer, String> consumer = createProducer();
+        consumer.send(new ProducerRecord<>("test-sasl-consumer", 1, "hi world"));
+        String string = RestAssured.when().get("/sasl").andReturn().asString();
+        Assertions.assertEquals("hi world", string);
+    }
+
+}

--- a/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/SslKafkaConsumerITCase.java
+++ b/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/SslKafkaConsumerITCase.java
@@ -3,6 +3,6 @@ package io.quarkus.it.kafka;
 import io.quarkus.test.junit.NativeImageTest;
 
 @NativeImageTest
-public class SaslKafkaConsumerITCase extends SaslKafkaConsumerTest {
+public class SslKafkaConsumerITCase extends SslKafkaConsumerTest {
 
 }

--- a/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/SslKafkaConsumerTest.java
+++ b/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/SslKafkaConsumerTest.java
@@ -21,12 +21,12 @@ import io.quarkus.test.junit.QuarkusTest;
 import io.restassured.RestAssured;
 
 @QuarkusTest
-@QuarkusTestResource(KafkaSASLTestResource.class)
-public class SaslKafkaConsumerTest {
+@QuarkusTestResource(KafkaSSLTestResource.class)
+public class SslKafkaConsumerTest {
 
     private static void addSsl(Properties props) {
         try {
-            File sslDir = KafkaSASLTestResource.sslDir(null, false);
+            File sslDir = KafkaSSLTestResource.sslDir(null, false);
             File tsFile = new File(sslDir, "kafka-truststore.p12");
             props.setProperty(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, "SSL");
             props.setProperty(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG, tsFile.getPath());
@@ -41,7 +41,7 @@ public class SaslKafkaConsumerTest {
     public static Producer<Integer, String> createProducer() {
         Properties props = new Properties();
         props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:19093");
-        props.put(ProducerConfig.CLIENT_ID_CONFIG, "sasl-test-producer");
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "test-ssl-producer");
         props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class.getName());
         props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
         addSsl(props);
@@ -53,7 +53,7 @@ public class SaslKafkaConsumerTest {
     public void testReception() {
         Producer<Integer, String> consumer = createProducer();
         consumer.send(new ProducerRecord<>("test-ssl-consumer", 1, "hi world"));
-        String string = RestAssured.when().get("/sasl").andReturn().asString();
+        String string = RestAssured.when().get("/ssl").andReturn().asString();
         Assertions.assertEquals("hi world", string);
     }
 

